### ISSUE TITLE
Destructuring an object literal has bogus excess property error and noImplicitAny error

### ISSUE THREAD

**Issue by sandersn** [Microsoft/TypeScript#8558]
**TypeScript Version:** 

At least 1.8 onward (tested in 1.8.10 and playground and master)

**Code**

```ts
var { x } = { a: 1, x: 12 };
```

**Expected behavior:**

No error and 'x' is of type number.

**Actual behavior:** 

```ts
var { x } = { a: 1, x: 12 };
              ~
Object literal may only specify known properties, and 'a' does not exist in '{ x: any }'.
```

If you add --noImplicitAny, you'll get another bogus error:

```ts
// @noImplicitAny: true
var { x } = { a: 1, x: 12 };
              ~
              (Same as above)
      ~
Binding element 'x' implicitly has 'any' type.
x./**/
```

But you can see that x is actually a number by looking at intellisense at the specified point.

It looks like contextual typing is messing with error reporting. Probably nobody actually writes no-op code like this, but our test suite has it in destructuringObjectBindingPatternAndAssignment3, with the above behaviour baselined as 'correct'.

**Comment by mhegazy** [comment]
i am inclined to say won't fix on this one.

**Comment by sandersn** [comment]
Yeah, it's do-nothing code. My object rest-spread PR already removes the test case as a by-product.
